src/main/java/io/camunda/testing/extensions/ZeebeAssertionsExtension.java

```diff
@@ -11,50 +11,66 @@
 import org.junit.jupiter.api.extension.ExtensionContext;
 import org.junit.platform.commons.util.ReflectionUtils;
 
-// TODO rewrite this mess
 public class ZeebeAssertionsExtension implements BeforeEachCallback, AfterEachCallback {
 
   private static final String KEY_ZEEBE_CLIENT = "ZEEBE_CLIENT";
 
   @Override
   public void beforeEach(final ExtensionContext extensionContext) throws Exception {
+    final Object testInstance = extensionContext.getRequiredTestInstance();
+    storeRecordStreamSource(extensionContext, testInstance);
+    storeZeebeClient(extensionContext, testInstance);
+  }
+
+  @Override
+  public void afterEach(final ExtensionContext extensionContext) {
+    RecordStreamSourceStore.reset();
+
+    final Object fieldContent = getStore(extensionContext).get(KEY_ZEEBE_CLIENT);
+    final ZeebeClient zeebeClient = (ZeebeClient) fieldContent;
+    zeebeClient.close();
+  }
+
+  private void storeRecordStreamSource(
+      final ExtensionContext extensionContext, final Object testInstance)
+      throws IllegalAccessException {
+    final Field recordStreamField = getRecordStreamField(extensionContext);
+    ReflectionUtils.makeAccessible(recordStreamField);
+    final RecordStreamSource recordStreamSource =
+        (RecordStreamSource) recordStreamField.get(testInstance);
+    RecordStreamSourceStore.init(recordStreamSource);
+  }
+
+  private Field getRecordStreamField(final ExtensionContext extensionContext) {
     final Optional<Field> recordStreamOptional =
-        Arrays.stream(extensionContext.getTestClass().get().getDeclaredFields())
+        Arrays.stream(extensionContext.getRequiredTestClass().getDeclaredFields())
```

Comment:
🙃  This change it not a refactoring, i.e. it changes the error message if the test class is missing for the extensionContext. TBH, I'm not sure whether that can actually happen, which brings me to something actually interesting.

🔧  There is something to say for dealing with the case that the class or the instance is missing (`getTestClass().orElseThrow(...)`), in which case we could help the user with a helpful error message similar to the missing required fields. All this depends on whether the class or the instance (below) can ever be missing. If they can never be missing than I prefer `getRequiredTestClass()`.

Reply:
I thought about this. I don't know in which scenario the test class could not be present. Maybe when using the extension on a non-testclass. But when that happens the `beforeEach()` wouldn't be executed in the first place.

I will stick with `getRequiredTestClass()` for now. If we ever run into this problem we can find a solution at that time.